crates/compiler/types/src/types.rs

```diff
@@ -333,6 +339,40 @@ impl OptAbleType {
     }
 }
 
+/// Polarity of a type, or roughly, what side of an arrow it appears on.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Polarity {
+    /// A type that appears in negative/input position
+    Neg,
+    /// A type that appears in positive/output position
+    Pos,
+}
```

Comment:
Our notion of "polarity" here is not exactly the standard notion of type polarity, but I still think it's the best name to use here because it generally aligns with descriptions in other type systems. I'd like to avoid an "input/output" enum here because this really is tracking polarity based on the side of an arrow, rather than contextually whether something is an input or output.